subsys/bluetooth/services/fast_pair/fp_gatt_service.c

```diff
@@ -281,8 +287,16 @@ static int parse_key_based_pairing_write(const struct bt_conn *conn,
 	}
 
 	if (memcmp(parsed_req->provider_address, &conn_info.le.local->a.val,
-	    sizeof(parsed_req->provider_address))) {
-		return -EINVAL;
+		   sizeof(parsed_req->provider_address))) {
+		/* This is a workaround of Android issue with sending old RPA address during
+		 * Key-based Pairing write. Normally, error value should be returned at this point.
+		 */
+		if (!bt_rpa_irk_matches(bt_dev.irk[conn_info.id],
+					(const bt_addr_t *)parsed_req->provider_address)) {
```

Comment:
The cast to `const bt_addr_t *` may be unsafe here. If at some point the structure gets modified, we may observe strange behavior on our end. Maybe we could at least assert that sizes are equal to prevent unexpected problems?